hubblestack/extmods/hubble_mods/time_sync.py

```diff
@@ -0,0 +1,223 @@
+
+FDG Example:
+------------
+main:
+  description: 'time_sync check'
+  module: time_sync
+    args:
+        ntp_servers:
+            - server1
+            - server2
+        extend_chained: true #optional
+
+Mandatory Params:
+    This module requires ntp_servers. That come either from args, or from chaining, or both
+'''
+
+
+import logging
+
+import hubblestack.extmods.module_runner.runner_utils as runner_utils
+from hubblestack.utils.hubble_error import HubbleCheckValidationError
+
+import salt.utils.platform
+
+if not salt.utils.platform.is_windows():
+    import ntplib
+log = logging.getLogger(__name__)
+
+def validate_params(block_id, block_dict, extra_args=None):
+    """
+    Validate all mandatory params required for this module
+
+    :param block_id:
+        id of the block
+    :param block_dict:
+        parameter for this module
+    :param extra_args:
+        Extra argument dictionary, (If any)
+        Example: {'chaining_args': {'result': ['server1', 'server2'], 'status': True},
+                  'caller': 'Audit'}
+
+    Raises:
+        AuditCheckValidationError: For any validation error
+    """
+    log.debug('Module: time_sync Start validating params for check-id: {0}'.format(block_id))
+
+    chain_args = None if not extra_args else extra_args.get('chaining_args')
+    ntp_servers = _get_ntp_servers(block_id, block_dict, chain_args)
+
+    if not ntp_servers:
+        raise HubbleCheckValidationError('No ntp_servers provided')
+
+    log.debug('Validation success for check-id: {0}'.format(block_id))
+
+def execute(block_id, block_dict, extra_args=None):
+    """
+    Function that queries a list of NTP servers and checks if the
+    offset is bigger than `max_offset` minutes. It expects the results from
+    at least `nb_servers` servers in the list, otherwise the check fails.
+
+    :param block_id:
+        id of the block
+    :param block_dict:
+        parameter for this module
+    :param extra_args:
+        Extra argument dictionary, (If any)
+        Example: {'chaining_args': {'result': ['server1', 'server2'], 'status': True},
+                  'caller': 'Audit'}
+
+    returns:
+        tuple of result(value) and status(boolean)
+    """
+    log.debug('Executing stat module for id: {0}'.format(block_id))
+
+    chain_args = None if not extra_args else extra_args.get('chaining_args')
+    ntp_servers = _get_ntp_servers(block_id, block_dict, chain_args)
+    
+    time_sync_result = []
+
+    for ntp_server in ntp_servers:
+        offset = _query_ntp_server(ntp_server)
+        if not offset:
+            time_sync_result.append({
+                'ntp_server': ntp_server,
+                'replied': False
+            })
+            continue
+
+        time_sync_result.append({
+            'ntp_server': ntp_server,
+            'replied': True,
+            'offset': offset
+        })
+
+    return runner_utils.prepare_positive_result_for_module(block_id, time_sync_result)
```

Comment:
If none of the servers replied, then we should return negative result.

Reply:
The purpose of a module is just to return the output. Its not the decision maker. Comparator will do this.
Its purpose is to return just the founded output.
Since its executed well, hence the positive result.

Reply:
Agreed but in that case we should prepare a negative result in case there is an exception while making a query to ntp server. We are returning None here only inn that case.